### PR TITLE
Fix UpdateBonePosition() breaking animations

### LINKED ISSUE #4739: Changing one player bone garbles the entire player model
The bones in the player model are not positioned correctly. When changing one bone, the whole player becomes garbled. After some experimentation, it seems that the code below will (approximately) put the player parts in the correct positions. Note that the values themselves are not even reasonable: e.g. to place the arm in the default position, it needs to be rotated about 8.5 degrees around one axis, and 180 degrees about another. Similarly, the default body position (angle 0°) makes its back side face forward!

```lua
player:set_bone_position("Body", {x=0, y=-3.3, z=0}, {x=0, y=180, z=0})
player:set_bone_position("Head", {x=0, y=6.75, z=0}, {x=0, y=180, z=0})
player:set_bone_position("Arm_Right", {x=-2, y=6.75, z=0}, {x=180, y=0, z=-8.5})
player:set_bone_position("Arm_Left", {x=2, y=6.75, z=0}, {x=180, y=0, z=8.5})
player:set_bone_position("Leg_Right", {x=-1, y=0, z=0}, {x=180, y=0, z=0})
player:set_bone_position("Leg_Left", {x=1, y=0, z=0}, {x=180, y=0, z=0})
```

The player before and after changing one single bone:
bones_0 bones_2

### PR DESCRIPTION
**Pull request by theviper121**
Add compact, short information about your PR for easier understanding:

- Fix UpdateBonePosition() breaking animations
- It animates joints after switching to EJUOR_CONTROL
- fixes #4739 and all its duplicates

## To do

This PR is **Ready for Review**.

## How to test
Set a bone position using set_bone_position on a animated model.

Im using this in a global step
player:set_bone_position("Head", {x = 0, y = 6.75, z = 0}, {x = -math.deg(player:get_look_vertical()), y = 0, z = 0})

### CODE REVIEW

**Comment by theviper121** [comment]
Frame blending also works when using EJUOR_CONTROL now

**Comment by LoneWolfHT** [comment]
If a client joins after the bone has been set the player with their bone set looks broken like in the PR.
If I set the bone again with the second player online that fixes it.

Should animations like the walk anim be playing when a bone is set and the player walks?

**Comment by LoneWolfHT** [comment]
Hmmm, what I do notice anim-related is that if I run this code below (If you use this code remember to change the name to your own instead of "Lone_Wolf")
```
//lua for i = 0, 10, 0.5 do minetest.after(i, function() minetest.get_player_by_name("Lone_Wolf"):set_bone_position("Head", {x=0, y=6.75, z=0}, {x=0, y=45, z=0}) end) end
```
the player will freeze at different stages of their current anim

(//lua is a Worldedit command)

**Comment by theviper121** [comment]
The player will appear to freeze on any client that does not have this patch. That is part of the issue this aims to fix.

**Comment by theviper121** [comment]
I was able to fix the issue you described LoneWolfHT. I believe this is proper now.

**Comment by LoneWolfHT** [comment]
Seems to be working better now. But if I use this command:
```
//lua minetest.get_player_by_name("Lone_Wolf"):set_bone_position("Head", {x=0, y=6.75, z=0}, {x=0, y=-45, z=0})
```
and trigger any player anim (Like walking or mining) I will see a brief flash of the broken model

**Comment by LoneWolfHT** [comment]
Would it be possible to add the option to make the bone position offset from the pos of the bone in its animation?
For example:
`player:set_bone_position("Head", {x=0, y=0, z=0}, {x=0, y=-45, z=0})`
would result in:
image

**Comment by theviper121** [comment]
It would be possible but would break any mods that rely on the current behavior

**Comment by theviper121** [comment]
Updated code again to prevent flashing when its applied to the local player.

**Comment by LoneWolfHT** [comment]
Works 100% clientside now.
But if I log on with another client that also has the PR it'll see the flashing when the main client starts an anim

**Comment by LoneWolfHT** [comment]
> It would be possible but would break any mods that rely on the current behavior

Could disable it by default maybe

**Comment by theviper121** [comment]
The flashing issue is the update from the player model not being rotated properly.

**Comment by theviper121** [comment]
This issue is not present when used on other models.

**Comment by LoneWolfHT** [comment]
> This issue is not present when used on other models.

Which ones?

**Comment by theviper121** [comment]
Any model that does not require bones to be rotated to be in there "home" position such as the player model we use on Viperian Realms. The flash you see is the bones render in the wrong position and then get rotated properly.

**Comment by rubenwardy** [comment]
The small size of this PR scares me greatly. I cannot be this simple

**Comment by theviper121** [comment]
I could make it more complicated if you like but this takes care of the issues. The only downside is that the player model in minetest_game needs to be updated to avoid a graphical glitch.

**Comment by sfan5** [comment]
The fact that models need to be updated *might* be a problem, since we're committed to backwards compatibility. Did `set_bone_position` work before with animated models?

**Comment by theviper121** [comment]
No before setting a single bone position or using set_bone_position with a animated model did not work.

**Comment by theviper121** [comment]
The player model only needs to be updated if you plan to make use of this new feature. Therefore backwards compatibility is maintained.

**Comment by ElCeejoYT** [comment]
It seems that if an animation has any bones that don't have a rotation different from the base rotation then they still get messed up.

**Comment by theviper121** [comment]
This has not been my experience ElCeejoYT. Could you provide example code and provide details on the model you are using?

**Comment by ElCeejoYT** [comment]
I don't think example code would help since I wasn't trying to do anything with the bones that went crazy. It seems like the case is that it effects bones that aren't parented to another bone.

**Comment by theviper121** [comment]
Could you provide a link to the model in question?

**Comment by ElCeejoYT** [comment]
Seems like github won't let me post .blends here. What would be a good way to send it to you?

**Comment by LoneWolfHT** [comment]
> What would be a good way to send it to you?

~~They could probably let you ssh in~~
#
I was fiddling with the player model in Blender recently and noticed that it actually is broken, the rotation seems to break if you set it to a nonzero val

File: src/client/content_cao.cpp [hunk 1]
```diff
@@ -890,6 +890,12 @@ void GenericCAO::updateNodePos()
 
 void GenericCAO::step(float dtime, ClientEnvironment *env)
 {
+	if (m_animated_meshnode)
+	{
```
Comment (nerzhul) [inline src/client/content_cao.cpp:894]: wrongly placed break

**Comment by theviper121** [comment]
I believe the issue with non rotated bones has been resolved.

File: src/client/content_cao.cpp [hunk 2]
```diff
@@ -893,6 +893,18 @@ void GenericCAO::step(float dtime, ClientEnvironment *env)
 	if (m_animated_meshnode){
 		m_animated_meshnode->animateJoints();
 		updateBonePosition();
+		//search through bones to find misplaced ones
+		for (u32 i = 0; i < m_animated_meshnode->getJointCount(); ++i) {
+			irr::scene::IBoneSceneNode* bone = m_animated_meshnode->getJointNode(i);
+			if (bone)
+			{
+				v3f bone_rot = bone->getRelativeTransformation().getRotationDegrees();
+				if (abs(bone_rot.X - bone->getRotation().X) > 179 && abs(bone_rot.X - bone->getRotation().X) < 181){ // Workaround for Irrlicht bug
```
Comment (sfan5) [inline src/client/content_cao.cpp:901]: This could use a comment explaining what is broken and how the workaround fixes it

Comment (theviper121) [reply src/client/content_cao.cpp]: Added a better description of the workaround code.

**Comment by theviper121** [comment]
The latest commit also addresses the old player model.

File: src/client/content_cao.cpp [hunk 3]
```diff
@@ -890,6 +890,40 @@ void GenericCAO::updateNodePos()
 
 void GenericCAO::step(float dtime, ClientEnvironment *env)
 {
+	if (m_animated_meshnode) {
+		m_animated_meshnode->animateJoints();
+		updateBonePosition();
+
+		// search through bones to find mistakenly rotated bones due to bug in Irrlicht
+		for (u32 i = 0; i < m_animated_meshnode->getJointCount(); ++i) {
```
Comment (SmallJoker) [inline src/client/content_cao.cpp:898]: I think this code would be better integrated into `updateBonePosition`

**Comment by oilboi** [comment]
Do it

**Comment by appgurueu** [comment]
> `player:set_bone_position("Head", {x=0, y=0, z=0}, {x=0, y=-45, z=0})`

I'd **love** to have something like this, it would greatly simplify many things. Currently mods like playeranim have to hard-code bone offsets (although they could also be determined by examining the model).

In order to not break compatibility, how about `player:get_model_bone_position` or something like that; @LoneWolfHT's example could then be written as

> `player:set_bone_position("Head", player:get_model_bone_position("Head"), {x=0, y=-45, z=0})`

**Comment by theviper121** [comment]
I think the requests to add relative bone positioning or the ability to retrieve bone positions from the models belongs in another PR. The aim of this PR is to provide bug fixes.

File: src/client/content_cao.cpp [hunk 4]
```diff
@@ -1360,16 +1365,41 @@ void GenericCAO::updateBonePosition()
...
-		std::string bone_name = (*ii).first;
-		v3f bone_pos = (*ii).second.X;
-		v3f bone_rot = (*ii).second.Y;
+	for (auto &it : m_bone_position) {
+		std::string bone_name = it.first;
 		irr::scene::IBoneSceneNode* bone = m_animated_meshnode->getJointNode(bone_name.c_str());
-		if(bone)
-		{
-			bone->setPosition(bone_pos);
+		if(bone){
```
Comment (sfan5) [inline src/client/content_cao.cpp:1371]:
```suggestion
		if (bone ){
```

File: src/client/content_cao.cpp [hunk 5]
```diff
@@ -1360,16 +1365,41 @@ void GenericCAO::updateBonePosition()
...
+		if (skip)
+			continue;
+
+		v3f bone_rot = bone->getRelativeTransformation().getRotationDegrees();
+		// Workaround for Irrlicht bug
+		// We check each bone to see if it has been rotated ~180deg from its expected position due to a bug in Irricht
+		// when using EJUOR_CONTROL joint control. If the bug is detected we update the bone the the proper position
+		// and update the bones transformation.
+		if (abs(bone_rot.X - bone->getRotation().X) > 179 &&
+				abs(bone_rot.X - bone->getRotation().X) < 181){
```
Comment (sfan5) [inline src/client/content_cao.cpp:1400]: I think the code would be better like this:
```c++
float offset = fabsf(bone->getRelativeTransformation().getRotationDegrees().X - bone->getRotation().X);
if (offset > 179.9f && offset < 180.1f) { 
```

Comment (theviper121) [reply src/client/content_cao.cpp:1400]: I don't believe this level of precision is required here but I will make the desired change.

**Comment by LoneWolfHT** [comment]
I tested again too with multiple clients, it all seems to work :+1:

**Pull request merged**